File: heron/lc015.java
public class Solution {

    private Set<List<Integer>> resultSet;

    public List<List<Integer>> threeSum(int[] nums) {

        resultSet = new HashSet<>();

        // First, sorts the input array
        Arrays.sort(nums);

        // Then, moves the iterator
        for (int i=0; i<nums.length-2; i++) {
            int l = i + 1;
            int r = nums.length - 1;

            while(l < r) {
                // Finally, we get the combinations
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == 0) {
                    // Saves the result here
                    saveResult(nums, i, l, r);
                    l++;
                    r--;
                } else if (sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }

        }

        // Converts the set to array
        List<List<Integer>> results = new ArrayList<>();
        results.addAll(resultSet);
        return results;

    }

    private void saveResult(int[] nums, int p1, int p2, int p3) {
        List<Integer> result = new ArrayList<>();
        result.add(nums[p1]);
        result.add(nums[p2]);
        result.add(nums[p3]);
        resultSet.add(result);
    }

}

File: alice/LC015.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        out = []
        
        if len(nums) < 3:
            return out

        nums.sort()

        for i in xrange(len(nums[:-2])):

            if i > 0 and nums[i] == nums[i - 1]:
                continue

            j = i + 1
            k = len(nums) - 1

            while k > j:
                temp = nums[i] + nums[j] + nums[k]
                if temp == 0:
                    out.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while k > j and nums[k] == nums[k - 1]:
                        k -= 1
                    j += 1
                    k -= 1
                elif temp > 0:
                    k -= 1
                else:
                    j += 1

        return out

s = Solution()
print s.threeSum([-1, 0, 1, 2, -1, 4])